src/v/cluster/log_eviction_stm.cc

```diff
@@ -275,34 +281,58 @@ ss::future<std::error_code> log_eviction_stm::replicate_command(
 }
 
 ss::future<> log_eviction_stm::apply(model::record_batch batch) {
+    if (batch.header().type != model::record_batch_type::prefix_truncate) {
+        co_return;
+    }
     /// The work done within apply() must be deterministic that way the start
     /// offset will always be the same value across all replicas
     ///
     /// Here all apply() does is move forward the in memory start offset, a
     /// background fiber is responsible for evicting as much as possible
-    if (unlikely(
-          batch.header().type == model::record_batch_type::prefix_truncate)) {
-        /// record_batches of type ::prefix_truncate are always of size 1
-        const auto truncate_offset = serde::from_iobuf<model::offset>(
-          batch.copy_records().begin()->release_key());
-        if (truncate_offset > _delete_records_eviction_offset) {
-            vlog(
-              _logger.info,
-              "Moving effective start offset to "
-              "truncate_point: {} last_applied: {} ntp: {}",
-              truncate_offset,
-              last_applied_offset(),
-              _raft->ntp());
+    /// record_batches of type ::prefix_truncate are always of size 1
+    const auto batch_type = serde::from_iobuf<uint8_t>(
+      batch.copy_records().begin()->release_key());
+    if (batch_type != prefix_truncate_key) {
+        vlog(
+          _logger.error,
+          "Unknown prefix_truncate batch type for {} at offset {}: {}",
+          _raft->ntp(),
+          batch.header().base_offset(),
+          batch_type);
+        co_return;
+    }
+    const auto record = serde::from_iobuf<prefix_truncate_record>(
+      batch.copy_records().begin()->release_value());
+    if (record.rp_start_offset == model::offset{}) {
+        // This may happen if the requested offset was not in the local log at
+        // time of replicating. We still need to have replicated it though so
+        // other STMs can honor it (e.g. archival).
+        vlog(
+          _logger.info,
+          "Replicated prefix_truncate batch for {} with no redpanda offset. "
```

Comment:
nit: I'd mention local `no local redpanda offset`

Reply:
Will address in a follow-up.